src/game.cpp

```diff
@@ -7382,7 +7382,7 @@ bool game::refill_vehicle_part(vehicle &veh, vehicle_part *part, bool test)
     const long fuel_per_charge = fuel_charges_to_amount_factor( ftype );
     long req = ceil( ( part->ammo_capacity() - part->ammo_remaining() ) / double( fuel_per_charge ) );
     long qty = std::min( req, avail );
-    part->ammo_set( ftype, part->ammo_remaining() + qty );
+    part->ammo_set( ftype, part->ammo_remaining() + ( qty * double( fuel_per_charge ) ) );
```

Comment:
The explicit cast seems useless. The result is converted to a `long` upon calling `ammo_set` anyway and that (converted) result will not be different (as compared to the original code). The input is all integers and there is no division, so the calculation will be exact.